### DISCUSSION TITLE
Apply or remove edit link conditionally

### DISCUSSION THREAD

**Discussion by TechAspirant** [serenity-is/Serenity#5773] [Help]
How can I apply or remove edit link based on row - cell value in grid ?

**Comment by minhhungit** [comment]
@TechAspirant you can use formatter, for example:

```javascript
getColumns() {
    let cols = super.getColumns();

    let idCol = Q.tryFirst(cols, x => x.field == StateRow.Fields.Id);
    if (idCol) {
        let oldFormat = idCol.format;

        idCol.format = (ctx) => {
            let item = ctx.item as StateRow;
            if (item.Id % 2 == 0) {
                return oldFormat(ctx);
            }
            else {
                return `<span class='badge badge-success'>${item.Id}</span>`;
            }
        }
    }

    return cols;
}
```

Above code will re-format Id cell, if Id is an odd number, it will show text instead a link

**Reply by TechAspirant** [reply]
It works, Thanks.
Can we add inline edit button in grid conditionally as above ? 
Edit button should be visible if Id is an odd number (referring above code) and on click edit dialog should get popup.
Thanks once again.

**Reply by minhhungit** [reply]
not sure what you mean, if you need edit link you can re-use oldFormat (above)
in case you need to open a difference dialog, you can return a button string with css class like inline-button example then handle it using onClick event

https://github.com/serenity-is/Serene/blob/master/Serene/Serene.Core/Modules/BasicSamples/Grids/InlineActionButtons/InlineActionGrid.ts

**Reply by TechAspirant** [reply]
I added the inline button but how can I hide the inline edit and delete button conditionally.
Below is my code
```
        getColumns() {
            let cols = super.getColumns();

            cols.unshift({
                field: 'Delete Row',
                name: '',
                format: ctx => '<a class="inline-action delete-row" title="delete">' + 
                    '<i class="fa fa-trash-o text-red"></i></a>',
                width: 24,
                minWidth: 24,
                maxWidth: 24                   
            });

            cols.splice(1, 0, {
                field: 'View Details',
                name: '',
                format: ctx => '<a class="inline-action edit-details" title="view details">' + 
                '<i class="fa fa-pencil-square-o text-blue"></i></a>',
                width: 24,
                minWidth: 24,
                maxWidth: 24 
            });

            var user = Administration.UserRow.getLookup().items.filter(
                item => item.Username == Authorization.userDefinition.Username
            );
            if (user.length > 0) {
                let idCol = Q.tryFirst(cols, x => x.field == AssetRow.Fields.AssetName);
                let editCol = Q.tryFirst(cols, x => x.field == 'Delete Row');
                let deleteCol = Q.tryFirst(cols, x => x.field == 'View Details');
                if (idCol && editCol && deleteCol) {
                    let oldFormat = idCol.format;
                    let editColOldFormat = editCol.format;
                    let deleteColOldFormat = deleteCol.format;

                    idCol.format = (ctx) => {
                        let item = ctx.item as AssetRow;
                        if (item.TenantId == user[0].TenantId) {
                            return oldFormat(ctx);
                        }
                        else {
                            return `<span>${item.AssetName}</span>`;
                        }
                    }
                }
            }

            return cols;
        }
```

**Reply by minhhungit** [reply]
Same with my example, here we will try to apply it with Delete button

```javascript
cols.unshift({
      field: 'Delete Row',
      name: '',
      format: ctx => {
          if (abc = xyz){ // your condition
               return '';
          }
         else {
               return '<a class="inline-action delete-row" title="delete">' + 
                    '<i class="fa fa-trash-o text-red"></i></a>
          }
      }',
      width: 24,
      minWidth: 24,
      maxWidth: 24                   
  });
```

**Comment by TechAspirant** [comment]
getColumns() {
            let cols = super.getColumns();

            var user = Administration.UserRow.getLookup().items.filter(
                item => item.Username == Authorization.userDefinition.Username
            );

            if (user.length > 0) {
                cols.unshift({
                    field: 'Delete Row',
                    name: '',
                    format: ctx => {
                        var link = '<a class="inline-action delete-row" title="delete">' +
                            '<i class="fa fa-trash-o text-red"></i></a>';
                        let item = ctx.item as AssetRow;
                        if (item.TenantId != user[0].TenantId) {
                            link = '<a class="inline-action delete-row-dummy" title="delete">' +
                                '<i class="fa fa-trash-o text-black"></i></a>';
                        }

                        return link;
                    },
                    width: 24,
                    minWidth: 24,
                    maxWidth: 24
                });

                cols.splice(1, 0, {
                    field: 'View Details',
                    name: '',
                    format: ctx => {
                        var link = '<a class="inline-action edit-details" title="view details">' +
                            '<i class="fa fa-pencil-square-o text-blue"></i></a>';
                        let item = ctx.item as AssetRow;
                        if (item.TenantId != user[0].TenantId) {
                            link = '<a class="inline-action edit-details-dummy" title="view details">' +
                                '<i class="fa fa-pencil-square-o text-black"></i></a>';
                        }

                        return link;
                    },
                    width: 24,
                    minWidth: 24,
                    maxWidth: 24
                });



                let idCol = Q.tryFirst(cols, x => x.field == AssetRow.Fields.AssetName);
                if (idCol) {
                    let oldFormat = idCol.format;
                    idCol.format = (ctx) => {
                        let item = ctx.item as AssetRow;
                        if (item.TenantId == user[0].TenantId) {
                            return oldFormat(ctx);
                        }
                        else {
                            return `<span>${item.AssetName}</span>`;
                        }
                    }
                }
            }

            return cols;
        }

This is how I managed. Is this right ? @minhhungit

**Reply by minhhungit** [reply]
yeah, and make sure that you also verify your logic (item.TenantId == user[0].TenantId) at serverside (in Repository Validate for example). Because hiding delete button at client side is not enough

**Reply by TechAspirant** [reply]
How can I validate whether at server side is validated or not ?

**Reply by minhhungit** [reply]
write code in repository

![image](https://user-images.githubusercontent.com/2279508/120989399-b3826600-c7a9-11eb-990d-05aaf85ddbd4.png)

**Reply by TechAspirant** [reply]
Is this right ?
```
        private class MyDeleteHandler : DeleteRequestHandler<MyRow> 
        {
            protected IUserAccessor UserAccessor { get; }
            protected IUserRetrieveService UserRetriever { get; }

            public MyDeleteHandler(IRequestContext context, IUserAccessor userAccessor, IUserRetrieveService userRetriever)
                 : base(context)
            {
                UserAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
                UserRetriever = userRetriever ?? throw new ArgumentNullException(nameof(userRetriever));
            }

            protected override void ValidateRequest()
            {
                base.ValidateRequest();

                if (IsUpdate)
                {
                    var username = UserAccessor.User?.Identity?.Name;

                    if (UserRetriever.ByUsername(username) is UserDefinition user)
                    {
                        if (Old.TenantId != user.TenantId)
                            Permissions.ValidatePermission(PermissionKeys.Tenants, Localizer);
                    }
                }
            }
        }
```

**Reply by minhhungit** [reply]
I don't think Delete has `IsUpdate`, also `Old` object
In delete handler I think it just has `Row` object, so just remove `IsUpdate` and change `Old` to `Row`

the rest looks good to me

**Reply by TechAspirant** [reply]
I am getting this error error CS0117: 'PermissionKeys' does not contain a definition for 'Tenants'

**Reply by TechAspirant** [reply]
This is how I managed for edit and delete
        private class MySaveHandler : SaveRequestHandler<MyRow>
        {
            protected IUserAccessor UserAccessor { get; }
            protected IUserRetrieveService UserRetriever { get; }

            public MySaveHandler(IRequestContext context, IUserAccessor userAccessor, IUserRetrieveService userRetriever)
                 : base(context)
            {
                UserAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
                UserRetriever = userRetriever ?? throw new ArgumentNullException(nameof(userRetriever));
            }

            protected override void ValidateRequest()
            {
                base.ValidateRequest();

                if (IsUpdate)
                {

                    var username = UserAccessor.User?.Identity?.Name;

                    if (UserRetriever.ByUsername(username) is UserDefinition user)
                    {
                        if (Fld.TenantId != user.TenantId)
                            throw new ValidationError("You cannot change or delete record.");
                    }
                }

            }
        }

        private class MyDeleteHandler : DeleteRequestHandler<MyRow>
        {
            protected IUserAccessor UserAccessor { get; }
            protected IUserRetrieveService UserRetriever { get; }

            public MyDeleteHandler(IRequestContext context, IUserAccessor userAccessor, IUserRetrieveService userRetriever)
                 : base(context)
            {
                UserAccessor = userAccessor ?? throw new ArgumentNullException(nameof(userAccessor));
                UserRetriever = userRetriever ?? throw new ArgumentNullException(nameof(userRetriever));
            }

            protected override void ValidateRequest()
            {
                base.ValidateRequest();

                var username = UserAccessor.User?.Identity?.Name;

                if (UserRetriever.ByUsername(username) is UserDefinition user)
                {
                    if (Fld.TenantId != user.TenantId)
                        throw new ValidationError("You cannot change or delete record.");
                }
            }
        }
